### PR TITLE
initial jag_conduit data reader

### PR DESCRIPTION
**Pull request by JaeseungYeom**
Initial implementation of the jag_conduit data reader, which load the jag data from the new conduit-packed HDF5 format file.

The capability of the current implementation includes:
- This reader handles the conduit-packed HDF5 JAG data file format which includes more scalar variables and input parameters than the numpy-based format.
- In addition, the new format includes multiple images, each of which is taken at one particular view at a particular time.
- The reader identifies the image taken at the time closest to the bang per view.
- The reader can be configured to select a particular set of scalar output variables as well as a set of input parameters in the data.
- Building LBANN with the reader is optional, and it is enabled by using --with-conduit option of the build script.

The plan for further improvements:
- Currently, no normalization is implemented.
- I plan to inherit this reader from data_reader_imagenet or data_reader_multi_images to make the existing image normalization functionality available.
- Currently, there is no hook to data_store. I will add them once inheritance structure is sorted out.
- Then I will hook this into prototext interface.
- In addition, I plan to create an offline mean image computation tool to allow data-set-wide normalization.
- Currently, building LBANN with the reader requires to rely on a pre-installed conduit library, which is linked with hdf5 libraries as well as other dependencies. Perhaps, someone can help me include the conduit building in the LBANN build.

### CODE REVIEW

File: include/lbann/data_readers/data_reader_jag_conduit.hpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+  /// Show the description
+  std::string get_description() const;
+
+  /// Return the image simulation output of the i-th sample
+  std::vector<cv::Mat> get_cv_images(const size_t i) const;
+
+  /**
+   * Return the images of the i-th sample as an 1-D vector of lbann::DataType
+   * There is one image per view, each of which is taken at closet to the bang time.
```
Comment (davidHysom) [inline include/lbann/data_readers/data_reader_jag_conduit.hpp]: closet -> closest ?

File: include/lbann/data_readers/data_reader_jag_conduit.hpp [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  void check_image_size();
+  /// Make sure that the keys to choose scalar outputs are valid
+  void check_scalar_keys();
+  /// Make sure that the keys to choose scalar outputs are valid
+  void check_input_keys();
+
+  /// Check if the given sample id is valid
+  bool check_sample_id(const size_t i) const;
+
+  /// Choose the image closet to the bang time among those associated with the i-th sample
```
Comment (davidHysom) [inline include/lbann/data_readers/data_reader_jag_conduit.hpp]: closet -> closest ?

File: include/lbann/data_readers/data_reader_jag_conduit.hpp [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  /// Return the image simulation output of the i-th sample
+  std::vector<cv::Mat> get_cv_images(const size_t i) const;
+
+  /**
+   * Return the images of the i-th sample as an 1-D vector of lbann::DataType
+   * There is one image per view, each of which is taken at closet to the bang time.
+   */
+  std::vector<ch_t> get_images(const size_t i) const;
+
+  /// Return the scalar simulatino output data of the i-th sample
```
Comment (davidHysom) [inline include/lbann/data_readers/data_reader_jag_conduit.hpp]: simulatino -> simulation

File: include/lbann/data_readers/data_reader_jag_conduit.hpp (hunk 1 above)
Comment (JaeseungYeom) [reply include/lbann/data_readers/data_reader_jag_conduit.hpp]: fixed

File: include/lbann/data_readers/data_reader_jag_conduit.hpp (hunk 2 above)
Comment (JaeseungYeom) [reply include/lbann/data_readers/data_reader_jag_conduit.hpp]: fixed

File: src/data_readers/data_reader_jag_conduit.cpp [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+
+#define _THROW_LBANN_EXCEPTION_(_CLASS_NAME_,_MSG_) { \
+  std::stringstream err; \
+  err << __FILE__ << ' '  << __LINE__ << " :: " \
+      << _CLASS_NAME_ << "::" << _MSG_; \
+  throw lbann_exception(err.str()); \
+}
+
+#define _THROW_EXCEPTION_(_MSG_) \
+        _THROW_LBANN_EXCEPTION_("data_reader_jag_conduit", _MSG_)
```
[the remark this answers is not in the document]
Comment (JaeseungYeom) [reply src/data_readers/data_reader_jag_conduit.cpp]: It is local to the scope of the file. We can probably define another macro for the class name, 
#define _CLASS_ "data_reader_jag_conduit"
and use it with _THROW_LBANN_EXCEPTION_(_CLASS_, _MSG_).
Then, put #undef _CLASS_ at the end of the file to make sure.

**Comment by JaeseungYeom** [comment]
Yes, this PR is missing any aspect of data_store.
While this do the simple task of loading data, it does not do any preprocessing. I plan to make this inherit from data_reader_multi_images or data_reader_imagenet in the next PR. Then, whether I can provide similar interfaces for data_store or not will become more clear.

**Pull request merged**